File: core/src/main/java/xyz/tynn/astring/AString.java
//  Copyright 2020 Christian Schmitz
//  SPDX-License-Identifier: Apache-2.0

package xyz.tynn.astring;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * A {@link String} abstraction for <i>Android</i>.
 *
 * The purpose of this type is to provide context sensitive strings.
 * This could be a plain {@link CharSequence} or a {@link android.text.Spannable},
 * a string resource or a formatted quantity string.
 *
 * {@code AString} is almost always used from the main thread,
 * therefore all implementations must be non-blocking since
 */
public interface AString {

    /**
     * Provides a context sensitive string
     *
     * @param context to access resources
     * @return string value from context; might be null
     */
    @Nullable
    CharSequence invoke(@NonNull Context context);
}

File: core/src/main/kotlin/xyz/tynn/astring/AString.kt
//  Copyright 2020 Christian Schmitz
//  SPDX-License-Identifier: Apache-2.0

package xyz.tynn.astring

import android.content.Context
import android.view.View
import androidx.fragment.app.Fragment
import kotlin.reflect.KProperty

/**
 * Invokes the `AString` with [Fragment.requireContext]
 */
operator fun AString.invoke(
    fragment: Fragment,
) = invoke(fragment.requireContext())

/**
 * Invokes the `AString` with [View.getContext]
 */
operator fun AString.invoke(
    view: View,
) = invoke(view.context)

/**
 * Delegates a `CharSequence?` property within a [Context] to the `AString`
 */
@JvmSynthetic
operator fun AString.getValue(
    thisRef: Context,
    property: KProperty<*>,
) = invoke(thisRef)

/**
 * Delegates a `CharSequence?` property within a [Fragment] to the `AString`
 */
@JvmSynthetic
operator fun AString.getValue(
    thisRef: Fragment,
    property: KProperty<*>,
) = invoke(thisRef.requireContext())

/**
 * Delegates a `CharSequence?` property within a [View] to the `AString`
 */
@JvmSynthetic
operator fun AString.getValue(
    thisRef: View,
    property: KProperty<*>,
) = invoke(thisRef.context)
